// @sentry/node/build/esm/integrations/tracing/graphql/vendored/enum.js
var AllowedOperationTypes = /* @__PURE__ */ ((AllowedOperationTypes2) => {
  AllowedOperationTypes2["QUERY"] = "query";
  AllowedOperationTypes2["MUTATION"] = "mutation";
  AllowedOperationTypes2["SUBSCRIPTION"] = "subscription";
  return AllowedOperationTypes2;
})(AllowedOperationTypes || {});
var TokenKind = /* @__PURE__ */ ((TokenKind2) => {
  TokenKind2["SOF"] = "<SOF>";
  TokenKind2["EOF"] = "<EOF>";
  TokenKind2["BANG"] = "!";
  TokenKind2["DOLLAR"] = "$";
  TokenKind2["AMP"] = "&";
  TokenKind2["PAREN_L"] = "(";
  TokenKind2["PAREN_R"] = ")";
  TokenKind2["SPREAD"] = "...";
  TokenKind2["COLON"] = ":";
  TokenKind2["EQUALS"] = "=";
  TokenKind2["AT"] = "@";
  TokenKind2["BRACKET_L"] = "[";
  TokenKind2["BRACKET_R"] = "]";
  TokenKind2["BRACE_L"] = "{";
  TokenKind2["PIPE"] = "|";
  TokenKind2["BRACE_R"] = "}";
  TokenKind2["NAME"] = "Name";
  TokenKind2["INT"] = "Int";
  TokenKind2["FLOAT"] = "Float";
  TokenKind2["STRING"] = "String";
  TokenKind2["BLOCK_STRING"] = "BlockString";
  TokenKind2["COMMENT"] = "Comment";
  return TokenKind2;
})(TokenKind || {});
var SpanNames = /* @__PURE__ */ ((SpanNames2) => {
  SpanNames2["EXECUTE"] = "graphql.execute";
  SpanNames2["PARSE"] = "graphql.parse";
  SpanNames2["RESOLVE"] = "graphql.resolve";
  SpanNames2["VALIDATE"] = "graphql.validate";
  SpanNames2["SCHEMA_VALIDATE"] = "graphql.validateSchema";
  SpanNames2["SCHEMA_PARSE"] = "graphql.parseSchema";
  return SpanNames2;
})(SpanNames || {});

export { AllowedOperationTypes, SpanNames, TokenKind };
