// @sentry/core/build/esm/integrations/mcp-server/attributes.js
const MCP_METHOD_NAME_ATTRIBUTE = "mcp.method.name";
const MCP_REQUEST_ID_ATTRIBUTE = "mcp.request.id";
const MCP_SESSION_ID_ATTRIBUTE = "mcp.session.id";
const MCP_TRANSPORT_ATTRIBUTE = "mcp.transport";
const MCP_SERVER_NAME_ATTRIBUTE = "mcp.server.name";
const MCP_SERVER_TITLE_ATTRIBUTE = "mcp.server.title";
const MCP_SERVER_VERSION_ATTRIBUTE = "mcp.server.version";
const MCP_PROTOCOL_VERSION_ATTRIBUTE = "mcp.protocol.version";
const MCP_TOOL_NAME_ATTRIBUTE = "mcp.tool.name";
const MCP_RESOURCE_URI_ATTRIBUTE = "mcp.resource.uri";
const MCP_PROMPT_NAME_ATTRIBUTE = "mcp.prompt.name";
const MCP_TOOL_RESULT_IS_ERROR_ATTRIBUTE = "mcp.tool.result.is_error";
const MCP_TOOL_RESULT_CONTENT_COUNT_ATTRIBUTE = "mcp.tool.result.content_count";
const MCP_PROMPT_RESULT_DESCRIPTION_ATTRIBUTE = "mcp.prompt.result.description";
const MCP_PROMPT_RESULT_MESSAGE_COUNT_ATTRIBUTE = "mcp.prompt.result.message_count";
const MCP_REQUEST_ARGUMENT = "mcp.request.argument";
const MCP_LOGGING_LEVEL_ATTRIBUTE = "mcp.logging.level";
const MCP_LOGGING_LOGGER_ATTRIBUTE = "mcp.logging.logger";
const MCP_LOGGING_DATA_TYPE_ATTRIBUTE = "mcp.logging.data_type";
const MCP_LOGGING_MESSAGE_ATTRIBUTE = "mcp.logging.message";
const NETWORK_TRANSPORT_ATTRIBUTE = "network.transport";
const NETWORK_PROTOCOL_VERSION_ATTRIBUTE = "network.protocol.version";
const CLIENT_ADDRESS_ATTRIBUTE = "client.address";
const CLIENT_PORT_ATTRIBUTE = "client.port";
const MCP_SERVER_OP_VALUE = "mcp.server";
const MCP_NOTIFICATION_CLIENT_TO_SERVER_OP_VALUE = "mcp.notification.client_to_server";
const MCP_NOTIFICATION_SERVER_TO_CLIENT_OP_VALUE = "mcp.notification.server_to_client";
const MCP_FUNCTION_ORIGIN_VALUE = "auto.function.mcp_server";
const MCP_NOTIFICATION_ORIGIN_VALUE = "auto.mcp.notification";
const MCP_ROUTE_SOURCE_VALUE = "route";

export { CLIENT_ADDRESS_ATTRIBUTE, CLIENT_PORT_ATTRIBUTE, MCP_FUNCTION_ORIGIN_VALUE, MCP_LOGGING_DATA_TYPE_ATTRIBUTE, MCP_LOGGING_LEVEL_ATTRIBUTE, MCP_LOGGING_LOGGER_ATTRIBUTE, MCP_LOGGING_MESSAGE_ATTRIBUTE, MCP_METHOD_NAME_ATTRIBUTE, MCP_NOTIFICATION_CLIENT_TO_SERVER_OP_VALUE, MCP_NOTIFICATION_ORIGIN_VALUE, MCP_NOTIFICATION_SERVER_TO_CLIENT_OP_VALUE, MCP_PROMPT_NAME_ATTRIBUTE, MCP_PROMPT_RESULT_DESCRIPTION_ATTRIBUTE, MCP_PROMPT_RESULT_MESSAGE_COUNT_ATTRIBUTE, MCP_PROTOCOL_VERSION_ATTRIBUTE, MCP_REQUEST_ARGUMENT, MCP_REQUEST_ID_ATTRIBUTE, MCP_RESOURCE_URI_ATTRIBUTE, MCP_ROUTE_SOURCE_VALUE, MCP_SERVER_NAME_ATTRIBUTE, MCP_SERVER_OP_VALUE, MCP_SERVER_TITLE_ATTRIBUTE, MCP_SERVER_VERSION_ATTRIBUTE, MCP_SESSION_ID_ATTRIBUTE, MCP_TOOL_NAME_ATTRIBUTE, MCP_TOOL_RESULT_CONTENT_COUNT_ATTRIBUTE, MCP_TOOL_RESULT_IS_ERROR_ATTRIBUTE, MCP_TRANSPORT_ATTRIBUTE, NETWORK_PROTOCOL_VERSION_ATTRIBUTE, NETWORK_TRANSPORT_ATTRIBUTE };
