// @sentry/opentelemetry/build/esm/index.browser.js
import { consoleSandbox } from '@sentry/core';
export { getClient, getDynamicSamplingContextFromSpan, shouldPropagateTraceForUrl, withStreamedSpan } from '@sentry/core';
export { S as SEMANTIC_ATTRIBUTE_SENTRY_GRAPHQL_OPERATION, b as SentryPropagator, c as SentrySampler, d as SentrySpanProcessor, f as continueTrace, g as enhanceDscWithOpenTelemetryRootSpanName, h as getActiveSpan, i as getRequestSpanData, j as getScopesFromContext, k as getSentryResource, l as getSpanKind, m as getTraceContextForScope, n as isSentryRequestSpan, o as openTelemetrySetupCheck, p as setOpenTelemetryContextAsyncContextStrategy, q as setupEventContextTrace, r as spanHasAttributes, t as spanHasEvents, u as spanHasKind, v as spanHasName, w as spanHasParentId, x as spanHasStatus, y as startInactiveSpan, z as startSpan, A as startSpanManual, B as suppressTracing, C as withActiveSpan, D as wrapClientClass, E as wrapContextManagerClass, F as wrapSamplingDecision } from './resource-CEl6olOs.js';
import '@opentelemetry/api';
import '@opentelemetry/semantic-conventions';
import '@opentelemetry/core';
import './debug-build-B98wrZ1j.js';
import '@opentelemetry/sdk-trace-base';

class SentryAsyncLocalStorageContextManager {
  constructor() {
    consoleSandbox(() => {
      console.error("SentryAsyncLocalStorageContextManager is not supported in the browser");
    });
  }
}

export { SentryAsyncLocalStorageContextManager };
