// @sentry/core/build/cjs/tracing/vercel-ai/vercel-ai-attributes.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const OPERATION_NAME_ATTRIBUTE = "operation.name";
const AI_OPERATION_ID_ATTRIBUTE = "ai.operationId";
const AI_PROMPT_ATTRIBUTE = "ai.prompt";
const AI_SCHEMA_ATTRIBUTE = "ai.schema";
const AI_RESPONSE_OBJECT_ATTRIBUTE = "ai.response.object";
const AI_VALUES_ATTRIBUTE = "ai.values";
const AI_RESPONSE_TEXT_ATTRIBUTE = "ai.response.text";
const AI_RESPONSE_TOOL_CALLS_ATTRIBUTE = "ai.response.toolCalls";
const AI_RESPONSE_FINISH_REASON_ATTRIBUTE = "ai.response.finishReason";
const AI_PROMPT_MESSAGES_ATTRIBUTE = "ai.prompt.messages";
const AI_PROMPT_TOOLS_ATTRIBUTE = "ai.prompt.tools";
const AI_MODEL_ID_ATTRIBUTE = "ai.model.id";
const AI_RESPONSE_PROVIDER_METADATA_ATTRIBUTE = "ai.response.providerMetadata";
const AI_USAGE_CACHED_INPUT_TOKENS_ATTRIBUTE = "ai.usage.cachedInputTokens";
const AI_USAGE_INPUT_TOKEN_DETAILS_ATTRIBUTE_PREFIX = "ai.usage.inputTokenDetails.";
const AI_TELEMETRY_FUNCTION_ID_ATTRIBUTE = "ai.telemetry.functionId";
const AI_USAGE_COMPLETION_TOKENS_ATTRIBUTE = "ai.usage.completionTokens";
const AI_USAGE_PROMPT_TOKENS_ATTRIBUTE = "ai.usage.promptTokens";
const AI_USAGE_TOKENS_ATTRIBUTE = "ai.usage.tokens";
const AI_TOOL_CALL_NAME_ATTRIBUTE = "ai.toolCall.name";
const AI_TOOL_CALL_ID_ATTRIBUTE = "ai.toolCall.id";
const AI_TOOL_CALL_ARGS_ATTRIBUTE = "ai.toolCall.args";
const AI_TOOL_CALL_RESULT_ATTRIBUTE = "ai.toolCall.result";

exports.AI_MODEL_ID_ATTRIBUTE = AI_MODEL_ID_ATTRIBUTE;
exports.AI_OPERATION_ID_ATTRIBUTE = AI_OPERATION_ID_ATTRIBUTE;
exports.AI_PROMPT_ATTRIBUTE = AI_PROMPT_ATTRIBUTE;
exports.AI_PROMPT_MESSAGES_ATTRIBUTE = AI_PROMPT_MESSAGES_ATTRIBUTE;
exports.AI_PROMPT_TOOLS_ATTRIBUTE = AI_PROMPT_TOOLS_ATTRIBUTE;
exports.AI_RESPONSE_FINISH_REASON_ATTRIBUTE = AI_RESPONSE_FINISH_REASON_ATTRIBUTE;
exports.AI_RESPONSE_OBJECT_ATTRIBUTE = AI_RESPONSE_OBJECT_ATTRIBUTE;
exports.AI_RESPONSE_PROVIDER_METADATA_ATTRIBUTE = AI_RESPONSE_PROVIDER_METADATA_ATTRIBUTE;
exports.AI_RESPONSE_TEXT_ATTRIBUTE = AI_RESPONSE_TEXT_ATTRIBUTE;
exports.AI_RESPONSE_TOOL_CALLS_ATTRIBUTE = AI_RESPONSE_TOOL_CALLS_ATTRIBUTE;
exports.AI_SCHEMA_ATTRIBUTE = AI_SCHEMA_ATTRIBUTE;
exports.AI_TELEMETRY_FUNCTION_ID_ATTRIBUTE = AI_TELEMETRY_FUNCTION_ID_ATTRIBUTE;
exports.AI_TOOL_CALL_ARGS_ATTRIBUTE = AI_TOOL_CALL_ARGS_ATTRIBUTE;
exports.AI_TOOL_CALL_ID_ATTRIBUTE = AI_TOOL_CALL_ID_ATTRIBUTE;
exports.AI_TOOL_CALL_NAME_ATTRIBUTE = AI_TOOL_CALL_NAME_ATTRIBUTE;
exports.AI_TOOL_CALL_RESULT_ATTRIBUTE = AI_TOOL_CALL_RESULT_ATTRIBUTE;
exports.AI_USAGE_CACHED_INPUT_TOKENS_ATTRIBUTE = AI_USAGE_CACHED_INPUT_TOKENS_ATTRIBUTE;
exports.AI_USAGE_COMPLETION_TOKENS_ATTRIBUTE = AI_USAGE_COMPLETION_TOKENS_ATTRIBUTE;
exports.AI_USAGE_INPUT_TOKEN_DETAILS_ATTRIBUTE_PREFIX = AI_USAGE_INPUT_TOKEN_DETAILS_ATTRIBUTE_PREFIX;
exports.AI_USAGE_PROMPT_TOKENS_ATTRIBUTE = AI_USAGE_PROMPT_TOKENS_ATTRIBUTE;
exports.AI_USAGE_TOKENS_ATTRIBUTE = AI_USAGE_TOKENS_ATTRIBUTE;
exports.AI_VALUES_ATTRIBUTE = AI_VALUES_ATTRIBUTE;
exports.OPERATION_NAME_ATTRIBUTE = OPERATION_NAME_ATTRIBUTE;
