// @sentry/node/build/cjs/integrations/tracing/mongo/vendored/semconv.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const ATTR_DB_CONNECTION_STRING = "db.connection_string";
const ATTR_DB_MONGODB_COLLECTION = "db.mongodb.collection";
const ATTR_DB_NAME = "db.name";
const ATTR_DB_OPERATION = "db.operation";
const ATTR_DB_STATEMENT = "db.statement";
const ATTR_DB_SYSTEM = "db.system";
const ATTR_NET_PEER_NAME = "net.peer.name";
const ATTR_NET_PEER_PORT = "net.peer.port";
const DB_SYSTEM_NAME_VALUE_MONGODB = "mongodb";
const DB_SYSTEM_VALUE_MONGODB = "mongodb";
const METRIC_DB_CLIENT_CONNECTIONS_USAGE = "db.client.connections.usage";

exports.ATTR_DB_CONNECTION_STRING = ATTR_DB_CONNECTION_STRING;
exports.ATTR_DB_MONGODB_COLLECTION = ATTR_DB_MONGODB_COLLECTION;
exports.ATTR_DB_NAME = ATTR_DB_NAME;
exports.ATTR_DB_OPERATION = ATTR_DB_OPERATION;
exports.ATTR_DB_STATEMENT = ATTR_DB_STATEMENT;
exports.ATTR_DB_SYSTEM = ATTR_DB_SYSTEM;
exports.ATTR_NET_PEER_NAME = ATTR_NET_PEER_NAME;
exports.ATTR_NET_PEER_PORT = ATTR_NET_PEER_PORT;
exports.DB_SYSTEM_NAME_VALUE_MONGODB = DB_SYSTEM_NAME_VALUE_MONGODB;
exports.DB_SYSTEM_VALUE_MONGODB = DB_SYSTEM_VALUE_MONGODB;
exports.METRIC_DB_CLIENT_CONNECTIONS_USAGE = METRIC_DB_CLIENT_CONNECTIONS_USAGE;
