// @sentry/node/build/esm/integrations/tracing/kafka/vendored/semconv.js
const ATTR_MESSAGING_BATCH_MESSAGE_COUNT = "messaging.batch.message_count";
const ATTR_MESSAGING_DESTINATION_NAME = "messaging.destination.name";
const ATTR_MESSAGING_DESTINATION_PARTITION_ID = "messaging.destination.partition.id";
const ATTR_MESSAGING_KAFKA_MESSAGE_KEY = "messaging.kafka.message.key";
const ATTR_MESSAGING_KAFKA_MESSAGE_TOMBSTONE = "messaging.kafka.message.tombstone";
const ATTR_MESSAGING_KAFKA_OFFSET = "messaging.kafka.offset";
const ATTR_MESSAGING_OPERATION_NAME = "messaging.operation.name";
const ATTR_MESSAGING_OPERATION_TYPE = "messaging.operation.type";
const ATTR_MESSAGING_SYSTEM = "messaging.system";
const MESSAGING_OPERATION_TYPE_VALUE_PROCESS = "process";
const MESSAGING_OPERATION_TYPE_VALUE_RECEIVE = "receive";
const MESSAGING_OPERATION_TYPE_VALUE_SEND = "send";
const MESSAGING_SYSTEM_VALUE_KAFKA = "kafka";
const METRIC_MESSAGING_CLIENT_CONSUMED_MESSAGES = "messaging.client.consumed.messages";
const METRIC_MESSAGING_CLIENT_OPERATION_DURATION = "messaging.client.operation.duration";
const METRIC_MESSAGING_CLIENT_SENT_MESSAGES = "messaging.client.sent.messages";
const METRIC_MESSAGING_PROCESS_DURATION = "messaging.process.duration";

export { ATTR_MESSAGING_BATCH_MESSAGE_COUNT, ATTR_MESSAGING_DESTINATION_NAME, ATTR_MESSAGING_DESTINATION_PARTITION_ID, ATTR_MESSAGING_KAFKA_MESSAGE_KEY, ATTR_MESSAGING_KAFKA_MESSAGE_TOMBSTONE, ATTR_MESSAGING_KAFKA_OFFSET, ATTR_MESSAGING_OPERATION_NAME, ATTR_MESSAGING_OPERATION_TYPE, ATTR_MESSAGING_SYSTEM, MESSAGING_OPERATION_TYPE_VALUE_PROCESS, MESSAGING_OPERATION_TYPE_VALUE_RECEIVE, MESSAGING_OPERATION_TYPE_VALUE_SEND, MESSAGING_SYSTEM_VALUE_KAFKA, METRIC_MESSAGING_CLIENT_CONSUMED_MESSAGES, METRIC_MESSAGING_CLIENT_OPERATION_DURATION, METRIC_MESSAGING_CLIENT_SENT_MESSAGES, METRIC_MESSAGING_PROCESS_DURATION };
