// @sentry/node/build/esm/integrations/tracing/postgres/vendored/semconv.js
const ATTR_DB_CLIENT_CONNECTION_POOL_NAME = "db.client.connection.pool.name";
const ATTR_DB_CLIENT_CONNECTION_STATE = "db.client.connection.state";
const ATTR_DB_CONNECTION_STRING = "db.connection_string";
const ATTR_DB_NAME = "db.name";
const ATTR_DB_STATEMENT = "db.statement";
const ATTR_DB_SYSTEM = "db.system";
const ATTR_DB_USER = "db.user";
const ATTR_NET_PEER_NAME = "net.peer.name";
const ATTR_NET_PEER_PORT = "net.peer.port";
const DB_CLIENT_CONNECTION_STATE_VALUE_IDLE = "idle";
const DB_CLIENT_CONNECTION_STATE_VALUE_USED = "used";
const DB_SYSTEM_VALUE_POSTGRESQL = "postgresql";
const METRIC_DB_CLIENT_CONNECTION_COUNT = "db.client.connection.count";
const METRIC_DB_CLIENT_CONNECTION_PENDING_REQUESTS = "db.client.connection.pending_requests";

export { ATTR_DB_CLIENT_CONNECTION_POOL_NAME, ATTR_DB_CLIENT_CONNECTION_STATE, ATTR_DB_CONNECTION_STRING, ATTR_DB_NAME, ATTR_DB_STATEMENT, ATTR_DB_SYSTEM, ATTR_DB_USER, ATTR_NET_PEER_NAME, ATTR_NET_PEER_PORT, DB_CLIENT_CONNECTION_STATE_VALUE_IDLE, DB_CLIENT_CONNECTION_STATE_VALUE_USED, DB_SYSTEM_VALUE_POSTGRESQL, METRIC_DB_CLIENT_CONNECTION_COUNT, METRIC_DB_CLIENT_CONNECTION_PENDING_REQUESTS };
