// @sentry/node/build/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const http = require('./integrations/http.js');
const index$n = require('./integrations/node-fetch/index.js');
const index$5 = require('./integrations/fs/index.js');
const express = require('./integrations/tracing/express.js');
const index$4 = require('./integrations/tracing/fastify/index.js');
const index$a = require('./integrations/tracing/graphql/index.js');
const index$d = require('./integrations/tracing/kafka/index.js');
const index$i = require('./integrations/tracing/lrumemoizer/index.js');
const index$j = require('./integrations/tracing/mongo/index.js');
const index$k = require('./integrations/tracing/mongoose/index.js');
const index$m = require('./integrations/tracing/mysql/index.js');
const index$l = require('./integrations/tracing/mysql2/index.js');
const index$r = require('./integrations/tracing/redis/index.js');
const index$p = require('./integrations/tracing/postgres/index.js');
const postgresjs = require('./integrations/tracing/postgresjs.js');
const index$q = require('./integrations/tracing/prisma/index.js');
const index$b = require('./integrations/tracing/hapi/index.js');
const index$c = require('./integrations/tracing/hono/index.js');
const index$f = require('./integrations/tracing/koa/index.js');
const index$2 = require('./integrations/tracing/connect/index.js');
const index$e = require('./integrations/tracing/knex/index.js');
const index$s = require('./integrations/tracing/tedious/index.js');
const index$6 = require('./integrations/tracing/genericPool/index.js');
const index$3 = require('./integrations/tracing/dataloader/index.js');
const index = require('./integrations/tracing/amqplib/index.js');
const index$t = require('./integrations/tracing/vercelai/index.js');
const index$o = require('./integrations/tracing/openai/index.js');
const index$1 = require('./integrations/tracing/anthropic-ai/index.js');
const index$9 = require('./integrations/tracing/google-genai/index.js');
const index$g = require('./integrations/tracing/langchain/index.js');
const index$h = require('./integrations/tracing/langgraph/index.js');
const launchDarkly = require('./integrations/featureFlagShims/launchDarkly.js');
const openFeature = require('./integrations/featureFlagShims/openFeature.js');
const statsig = require('./integrations/featureFlagShims/statsig.js');
const unleash = require('./integrations/featureFlagShims/unleash.js');
const growthbook = require('./integrations/featureFlagShims/growthbook.js');
const firebase = require('./integrations/tracing/firebase/firebase.js');
const index$8 = require('./sdk/index.js');
const initOtel = require('./sdk/initOtel.js');
const index$7 = require('./integrations/tracing/index.js');
const opentelemetry = require('@sentry/opentelemetry');
const core = require('@sentry/core');
const nodeCore = require('@sentry/node-core');



exports.httpIntegration = http.httpIntegration;
exports.nativeNodeFetchIntegration = index$n.nativeNodeFetchIntegration;
exports.fsIntegration = index$5.fsIntegration;
exports.expressIntegration = express.expressIntegration;
exports.setupExpressErrorHandler = express.setupExpressErrorHandler;
exports.fastifyIntegration = index$4.fastifyIntegration;
exports.setupFastifyErrorHandler = index$4.setupFastifyErrorHandler;
exports.graphqlIntegration = index$a.graphqlIntegration;
exports.kafkaIntegration = index$d.kafkaIntegration;
exports.lruMemoizerIntegration = index$i.lruMemoizerIntegration;
exports.mongoIntegration = index$j.mongoIntegration;
exports.mongooseIntegration = index$k.mongooseIntegration;
exports.mysqlIntegration = index$m.mysqlIntegration;
exports.mysql2Integration = index$l.mysql2Integration;
exports.redisIntegration = index$r.redisIntegration;
exports.postgresIntegration = index$p.postgresIntegration;
exports.postgresJsIntegration = postgresjs.postgresJsIntegration;
exports.prismaIntegration = index$q.prismaIntegration;
exports.hapiIntegration = index$b.hapiIntegration;
exports.setupHapiErrorHandler = index$b.setupHapiErrorHandler;
exports.honoIntegration = index$c.honoIntegration;
exports.setupHonoErrorHandler = index$c.setupHonoErrorHandler;
exports.koaIntegration = index$f.koaIntegration;
exports.setupKoaErrorHandler = index$f.setupKoaErrorHandler;
exports.connectIntegration = index$2.connectIntegration;
exports.setupConnectErrorHandler = index$2.setupConnectErrorHandler;
exports.knexIntegration = index$e.knexIntegration;
exports.tediousIntegration = index$s.tediousIntegration;
exports.genericPoolIntegration = index$6.genericPoolIntegration;
exports.dataloaderIntegration = index$3.dataloaderIntegration;
exports.amqplibIntegration = index.amqplibIntegration;
exports.vercelAIIntegration = index$t.vercelAIIntegration;
exports.openAIIntegration = index$o.openAIIntegration;
exports.anthropicAIIntegration = index$1.anthropicAIIntegration;
exports.googleGenAIIntegration = index$9.googleGenAIIntegration;
exports.langChainIntegration = index$g.langChainIntegration;
exports.langGraphIntegration = index$h.langGraphIntegration;
exports.buildLaunchDarklyFlagUsedHandler = launchDarkly.buildLaunchDarklyFlagUsedHandlerShim;
exports.launchDarklyIntegration = launchDarkly.launchDarklyIntegrationShim;
exports.OpenFeatureIntegrationHook = openFeature.OpenFeatureIntegrationHookShim;
exports.openFeatureIntegration = openFeature.openFeatureIntegrationShim;
exports.statsigIntegration = statsig.statsigIntegrationShim;
exports.unleashIntegration = unleash.unleashIntegrationShim;
exports.growthbookIntegration = growthbook.growthbookIntegrationShim;
exports.firebaseIntegration = firebase.firebaseIntegration;
exports.getDefaultIntegrations = index$8.getDefaultIntegrations;
exports.getDefaultIntegrationsWithoutPerformance = index$8.getDefaultIntegrationsWithoutPerformance;
exports.init = index$8.init;
exports.initWithoutDefaultIntegrations = index$8.initWithoutDefaultIntegrations;
exports.initOpenTelemetry = initOtel.initOpenTelemetry;
exports.preloadOpenTelemetry = initOtel.preloadOpenTelemetry;
exports.getAutoPerformanceIntegrations = index$7.getAutoPerformanceIntegrations;
exports.setNodeAsyncContextStrategy = opentelemetry.setOpenTelemetryContextAsyncContextStrategy;
exports.SDK_VERSION = core.SDK_VERSION;
exports.SEMANTIC_ATTRIBUTE_SENTRY_OP = core.SEMANTIC_ATTRIBUTE_SENTRY_OP;
exports.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN = core.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE = core.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE = core.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE;
exports.Scope = core.Scope;
exports.addBreadcrumb = core.addBreadcrumb;
exports.addEventProcessor = core.addEventProcessor;
exports.addIntegration = core.addIntegration;
exports.captureCheckIn = core.captureCheckIn;
exports.captureConsoleIntegration = core.captureConsoleIntegration;
exports.captureEvent = core.captureEvent;
exports.captureException = core.captureException;
exports.captureFeedback = core.captureFeedback;
exports.captureMessage = core.captureMessage;
exports.captureSession = core.captureSession;
exports.close = core.close;
exports.consoleLoggingIntegration = core.consoleLoggingIntegration;
exports.continueTrace = core.continueTrace;
exports.createConsolaReporter = core.createConsolaReporter;
exports.createLangChainCallbackHandler = core.createLangChainCallbackHandler;
exports.createTransport = core.createTransport;
exports.dedupeIntegration = core.dedupeIntegration;
exports.endSession = core.endSession;
exports.eventFiltersIntegration = core.eventFiltersIntegration;
exports.expressErrorHandler = core.expressErrorHandler;
exports.extraErrorDataIntegration = core.extraErrorDataIntegration;
exports.featureFlagsIntegration = core.featureFlagsIntegration;
exports.flush = core.flush;
exports.functionToStringIntegration = core.functionToStringIntegration;
exports.getActiveSpan = core.getActiveSpan;
exports.getClient = core.getClient;
exports.getCurrentScope = core.getCurrentScope;
exports.getGlobalScope = core.getGlobalScope;
exports.getIsolationScope = core.getIsolationScope;
exports.getRootSpan = core.getRootSpan;
exports.getSpanDescendants = core.getSpanDescendants;
exports.getSpanStatusFromHttpCode = core.getSpanStatusFromHttpCode;
exports.getTraceData = core.getTraceData;
exports.getTraceMetaTags = core.getTraceMetaTags;
exports.httpHeadersToSpanAttributes = core.httpHeadersToSpanAttributes;
exports.inboundFiltersIntegration = core.inboundFiltersIntegration;
exports.instrumentAnthropicAiClient = core.instrumentAnthropicAiClient;
exports.instrumentGoogleGenAIClient = core.instrumentGoogleGenAIClient;
exports.instrumentLangChainEmbeddings = core.instrumentLangChainEmbeddings;
exports.instrumentLangGraph = core.instrumentLangGraph;
exports.instrumentOpenAiClient = core.instrumentOpenAiClient;
exports.instrumentStateGraphCompile = core.instrumentStateGraphCompile;
exports.instrumentSupabaseClient = core.instrumentSupabaseClient;
exports.isEnabled = core.isEnabled;
exports.isInitialized = core.isInitialized;
exports.lastEventId = core.lastEventId;
exports.linkedErrorsIntegration = core.linkedErrorsIntegration;
exports.parameterize = core.parameterize;
exports.profiler = core.profiler;
exports.requestDataIntegration = core.requestDataIntegration;
exports.rewriteFramesIntegration = core.rewriteFramesIntegration;
exports.setContext = core.setContext;
exports.setConversationId = core.setConversationId;
exports.setCurrentClient = core.setCurrentClient;
exports.setExtra = core.setExtra;
exports.setExtras = core.setExtras;
exports.setHttpStatus = core.setHttpStatus;
exports.setMeasurement = core.setMeasurement;
exports.setTag = core.setTag;
exports.setTags = core.setTags;
exports.setUser = core.setUser;
exports.spanStreamingIntegration = core.spanStreamingIntegration;
exports.spanToBaggageHeader = core.spanToBaggageHeader;
exports.spanToJSON = core.spanToJSON;
exports.spanToTraceHeader = core.spanToTraceHeader;
exports.startInactiveSpan = core.startInactiveSpan;
exports.startNewTrace = core.startNewTrace;
exports.startSession = core.startSession;
exports.startSpan = core.startSpan;
exports.startSpanManual = core.startSpanManual;
exports.supabaseIntegration = core.supabaseIntegration;
exports.suppressTracing = core.suppressTracing;
exports.trpcMiddleware = core.trpcMiddleware;
exports.updateSpanName = core.updateSpanName;
exports.winterCGHeadersToDict = core.winterCGHeadersToDict;
exports.withActiveSpan = core.withActiveSpan;
exports.withIsolationScope = core.withIsolationScope;
exports.withMonitor = core.withMonitor;
exports.withScope = core.withScope;
exports.wrapMcpServerWithSentry = core.wrapMcpServerWithSentry;
exports.zodErrorsIntegration = core.zodErrorsIntegration;
exports.NODE_VERSION = nodeCore.NODE_VERSION;
exports.NodeClient = nodeCore.NodeClient;
exports.SentryContextManager = nodeCore.SentryContextManager;
exports._INTERNAL_normalizeCollectionInterval = nodeCore._INTERNAL_normalizeCollectionInterval;
exports.anrIntegration = nodeCore.anrIntegration;
exports.childProcessIntegration = nodeCore.childProcessIntegration;
exports.consoleIntegration = nodeCore.consoleIntegration;
exports.contextLinesIntegration = nodeCore.contextLinesIntegration;
exports.createGetModuleFromFilename = nodeCore.createGetModuleFromFilename;
exports.createSentryWinstonTransport = nodeCore.createSentryWinstonTransport;
exports.cron = nodeCore.cron;
exports.defaultStackParser = nodeCore.defaultStackParser;
exports.disableAnrDetectionForCallback = nodeCore.disableAnrDetectionForCallback;
exports.generateInstrumentOnce = nodeCore.generateInstrumentOnce;
exports.getSentryRelease = nodeCore.getSentryRelease;
exports.httpServerIntegration = nodeCore.httpServerIntegration;
exports.httpServerSpansIntegration = nodeCore.httpServerSpansIntegration;
exports.localVariablesIntegration = nodeCore.localVariablesIntegration;
exports.logger = nodeCore.logger;
exports.makeNodeTransport = nodeCore.makeNodeTransport;
exports.metrics = nodeCore.metrics;
exports.modulesIntegration = nodeCore.modulesIntegration;
exports.nodeContextIntegration = nodeCore.nodeContextIntegration;
exports.nodeRuntimeMetricsIntegration = nodeCore.nodeRuntimeMetricsIntegration;
exports.onUncaughtExceptionIntegration = nodeCore.onUncaughtExceptionIntegration;
exports.onUnhandledRejectionIntegration = nodeCore.onUnhandledRejectionIntegration;
exports.pinoIntegration = nodeCore.pinoIntegration;
exports.processSessionIntegration = nodeCore.processSessionIntegration;
exports.spotlightIntegration = nodeCore.spotlightIntegration;
exports.systemErrorIntegration = nodeCore.systemErrorIntegration;
exports.validateOpenTelemetrySetup = nodeCore.validateOpenTelemetrySetup;
exports.withStreamedSpan = nodeCore.withStreamedSpan;
